caffe2/ideep/operators/quantization/int8_pool_op.cc

```diff
@@ -0,0 +1,63 @@
+#include <caffe2/ideep/operators/conv_pool_base_op.h>
+
+using namespace caffe2;
+
+namespace {
+
+class IDEEPInt8PoolOp final : public IDEEPConvPoolOpBase {
+ public:
+  USE_IDEEP_DEF_ALIASES();
+  USE_IDEEP_CONV_POOL_BASE_FUNCTIONS();
+
+  IDEEPInt8PoolOp(const OperatorDef& operator_def, Workspace* ws)
+      : IDEEPConvPoolOpBase(operator_def, ws) {
+    CAFFE_ENFORCE(
+        (dilation_h() == 1) && (dilation_w() == 1),
+        "Pooling op does not support dilation right now.");
+    if (!global_pooling_) {
+      CAFFE_ENFORCE(
+          pad_t() < kernel_h() && pad_b() < kernel_h() &&
+              pad_l() < kernel_w() && pad_r() < kernel_w(),
+          "Pad should be smaller than kernel.");
+    }
+
+    // Figure out the pooling descriptor.
+    if (operator_def.type().substr(0, 11) == "Int8MaxPool") {
+      algo_ = ialgo::pooling_max;
+    } else if (operator_def.type().substr(0, 15) == "Int8AveragePool") {
+      algo_ = ialgo::pooling_avg;
+    } else {
+      LOG(FATAL) << "Unsupported pooling method: " << operator_def.type();
+    }
+  }
+  ~IDEEPInt8PoolOp() override {}
+
+  bool RunOnDeviceWithOrderNCHW() override {
+    auto& X = Input(INPUT);
+    auto* Y = Output(OUTPUT);
+    auto Y_dims = CalcOutputDims(X, X.get_dim(1));
+
+    if (cached_X_descriptor_ != X.get_descriptor()) {
+      op_key_.clear();
+      cached_X_descriptor_ = X.dup_descriptor();
+    }
+
+    ideep::pooling_forward::compute(op_key_, X, Y_dims, *Y,
+        stride_, kernel_, pad_tl(), pad_br(), algo_, iprop::forward_inference);
+
+    return true;
+  }
+
+ private:
+  ialgo algo_;
+  ikey op_key_;
+  itensor::descriptor cached_X_descriptor_;
+
+  INPUT_TAGS(INPUT);
+  OUTPUT_TAGS(OUTPUT);
+};
+
+REGISTER_IDEEP_OPERATOR_WITH_ENGINE(Int8MaxPool, DNNLOWP, IDEEPInt8PoolOp);
```

Comment:
How does this co-exist with `fbgemm` based DNNLOP int8 engine?

Reply:
same engine but different device.

Reply:
I see.